apps/cms.c

```diff
@@ -885,10 +892,29 @@ int cms_main(int argc, char **argv)
             goto end;
     }
 
-    in = bio_open_default(infile, 'r',
-                          binary_files ? FORMAT_BINARY : informat);
-    if (in == NULL)
-        goto end;
+    if (operation != SMIME_SIGN && digesthex != NULL) {
+        BIO_printf(bio_err,
+                   "Warning: -digest is ignored for non-signing operation\n");
+        digesthex = NULL;
+    }
```

Comment:
I am sorry, but I do not see the point in allowing and just warning with this particular option combination.

Reply:
I don't care if it's an error or a warning, but I think you need to decide with @DDvO before I change this again. (I don't know which decision model you're using (raft consensus algorithm or foundation wins over corporation?))

Reply:
@DDvO We explicitly disallow using -digest with -in or -nodetach. Why we would allow using -digest (and ignoring it) with encryption or any other operation than -sign?

Usually we did just warnings and no errors in cases where the operation silently ignored some option before and we wanted to explicitly show that it is ignored. I do not see a point in such warnings otherwise if the combination of options is clearly invalid.

Reply:
One more case to ease use of a command would be to add just warning for options that are usually added to all the command invocations with some exceptional ones. In that case just warning and no error might make also sense. But I do not see the -digest being such option - it really does not have any sense with any other operation than signing.

Reply:
As I tried to point out yesterday in https://github.com/openssl/openssl/pull/15348#discussion_r839854201, an inconsistency between options can be more or less severe.

For instance, using -digest along with -in (and likewise -nodetach) appears to indicate that the user did not understand that (s)he asked for something impossible, namely to have both forms of hash input at the same time, so I'd go for an error here.

On the other hand, sometimes it is convenient if an app glossed over needless parameters (while ideally giving a warning):
Suppose a user calls `openssl cms` several times, e.g., changing forth and back between the -sign and -verify use cases (while possibly adapting some more options).
Then it helps not having to remove from the previous invocation all those options that are currently not needed but are not harmful.
Therefore, for such "mild" forms of inconsistencies, I'd go for a warning, not for an error.

Reply:
Also when adding checks to other apps (cmp, x509, req, etc.), in a nutshell my guiding strategy has been this:
* throw an error if the users asked for something impossible/contradictory
* give a warning if the user specified something irrelevant

Though admittedly this distinction can be more or less blurry.

Reply:
I am sorry but I do not really see -digest parameter as something that should be handled this way.

Reply:
All right, I do not insist in having no error in case -digest is needlessly given.